core/src/main/java/org/apache/iceberg/BaseTable.java

```diff
@@ -59,6 +60,45 @@ public Schema schema() {
     return ops.current().schema();
   }
 
+  @Override
+  public Schema schemaForSnapshot(long snapshotId) {
+    TableMetadata current = ops.current();
+    if (current.currentSnapshot() != null &&
+        current.currentSnapshot().snapshotId() == snapshotId) {
+      return current.schema();
+    }
+    Long snapshotTs = null;
+    for (HistoryEntry logEntry : current.snapshotLog()) {
+      if (logEntry.snapshotId() == snapshotId) {
+        snapshotTs = logEntry.timestampMillis();
+      }
+    }
+    Preconditions.checkArgument(snapshotTs != null,
+        "Cannot find a snapshot with id %s", snapshotId);
+    TableMetadata.MetadataLogEntry metadataLogEntry = null;
+    for (TableMetadata.MetadataLogEntry logEntry : current.previousFiles()) {
+      if (logEntry.timestampMillis() <= snapshotTs) {
+        metadataLogEntry = logEntry;
+      }
+    }
+    String metadataFile = metadataLogEntry.file();
+    TableMetadata metadata = TableMetadataParser.read(io(), metadataFile);
+    return metadata.schema();
+  }
+
+  @Override
+  public Schema schemaForSnapshotAsOfTime(long timestampMillis) {
```

Comment:
We don't need two methods. We can use the history log to find out what snapshot was current at a given time, then get the schema from that snapshot.

Reply:
Isn't that what I'm doing here? This method can be considered a convenience method.

Reply:
Yes, the concern is that we want to keep the `Table` API small and focused.

Reply:
I have kept `schemaForSnapshotAsOfTime` for convenience. Since the `Table` API is unchanged, I think it is ok.